### PR TITLE
feat: Implement reconstruct wallet seed flow

### PR DESCRIPTION
**Pull request by kartikCypherock**
Implements https://app.clickup.com/t/9002019994/PRF-4798

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4798 Task controller merge - Reconstruct private key](https://app.clickup.com/t/9002019994/PRF-4798)

File: src/wallet/reconstruct_seed_flow.c [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+reconstruct_state_e reconstruct_seed_handler(reconstruct_state_e state,
+                                             uint8_t *seed_out) {
+  reconstruct_state_e next_state = EXIT;
+  switch (state) {
+    case PASSPHRASE_INPUT: {
+      if (!WALLET_IS_PASSPHRASE_SET(wallet.wallet_info)) {
```
Comment (ujjwal-cyph) [inline src/wallet/reconstruct_seed_flow.c:134]: do reset/clear `wallet_credential_data.passphrase` before this check, for sanity.

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]

File: src/wallet/reconstruct_seed_flow.c [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+      memzero(wallet_shamir_data.mnemonic_shares,
+              sizeof(wallet_shamir_data.mnemonic_shares));
+      mnemonic_clear();
+      const char *mnemo =
+          mnemonic_from_data(secret, wallet.number_of_mnemonics * 4 / 3);
+
+      ASSERT(mnemo != NULL);
+      uint8_t seed[64] = {0};
+
+      mnemonic_to_seed(mnemo, wallet_credential_data.passphrase, seed, NULL);
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.c:235]: Correct, I have updated

File: src/wallet/reconstruct_seed_flow.c [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  }
+
+  // TODO: Consolidate in one function
+  // Clear confidential data irrespective of the result of the flow
+  memzero(&wallet, sizeof(wallet));
+  memzero(&wallet_shamir_data, sizeof(wallet_shamir_data));
+  memzero(&wallet_credential_data, sizeof(wallet_credential_data));
+
+  // Select wallet based on wallet_id
+  wallet_selector(wallet_id, &wallet);
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.c:269]: We should check the return code, I have added the check and mitigation

File: src/wallet/wallet_list.h [hunk 4]
```diff
@@ -41,4 +41,32 @@
...
+ * and is in VALID_WALLET state.
+ * @return false If the wallet corresponding to wallet_id is not found or is not
+ * in VALID_WALLET state.
+ */
+bool wallet_selector(const uint8_t *wallet_id, Wallet *wallet);
+
+/**
+ * @brief This API searches for wallet on the flash using wallet_id as key
+ * @details If a VALID_WALLET is found in the flash, then this API fills the
+ * wallet_name buffer
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply src/wallet/wallet_list.h:62]: Updated

File: src/wallet/wallet_list.h [hunk 5]
```diff
@@ -41,4 +41,32 @@
...
+ *
+ * @param wallet_id The wallet_id that needs to be searched
+ * @param wallet_name The buffer in which wallet name will be filled
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply src/wallet/wallet_list.h:65]: Updated

File: src/wallet/reconstruct_seed_flow.c [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+/*****************************************************************************
+ * EXTERN VARIABLES
+ *****************************************************************************/
+extern char *ALPHABET;
+extern char *ALPHA_NUMERIC;
+extern char *NUMBERS;
+extern char *PASSPHRASE;
+extern Wallet_shamir_data wallet_shamir_data;
+extern Wallet_credential_data wallet_credential_data;
+extern Wallet wallet;
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.c:81]: Added

File: src/wallet/reconstruct_seed_flow.c [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+    case TAP_CARD_FLOW: {
+      card_error_type_e card_status = card_flow_reconstruct_wallet(1);
+
+      if (CARD_OPERATION_SUCCESS == card_status) {
+        next_state = RECONSTRUCT_SEED;
+      } else if (CARD_OPERATION_INCORRECT_PIN_ENTERED == card_status) {
+        next_state = PIN_INPUT;
+      } else {
+        next_state = COMPLETED_WITH_ERRORS;
+      }
```
[the remark this answers is not in the document]
Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.c:201]: I have listed the cases which will result in the else clause

**Review by ujjwal-cyph — APPROVED** [review]

File: src/wallet/wallet_list.h [hunk 8]
```diff
@@ -41,4 +41,33 @@
...
+ *
+ * @param wallet_id The wallet_id that needs to be searched
+ * @param wallet Reference to Wallet structure which will be populated by the
+ * function
+ * @return true If the wallet corresponding to wallet_id is found in the flash
+ * and is in VALID_WALLET state.
+ * @return false If the wallet corresponding to wallet_id is not found or is not
+ * in VALID_WALLET state.
+ */
+bool wallet_selector(const uint8_t *wallet_id, Wallet *wallet);
```
Comment (amanCypherock) [inline src/wallet/wallet_list.h:57]: `get_valid_wallet_data` seems like a better name for this function as it actually fetches relevant data for a valid wallet.

File: src/wallet/wallet_list.h [hunk 9]
```diff
@@ -41,4 +41,33 @@
...
+ *
+ * @param wallet_id The wallet_id that needs to be searched
+ * @param wallet_name The buffer in which wallet name will be filled or NULL if
+ * wallet name is not required
+ * @return true If the wallet corresponding to wallet_id is found in the flash
+ * and is in VALID_WALLET state.
+ * @return false If the wallet corresponding to wallet_id is not found or is not
+ * in VALID_WALLET state.
+ */
+bool get_wallet_data(const uint8_t *wallet_id, uint8_t *wallet_name);
```
Comment (amanCypherock) [inline src/wallet/wallet_list.h:72]: The name is confusing to me, we could rename as `find_valid_wallet`.

This function is not being used right now, What would be the intended usage of this function

File: src/wallet/wallet_list.h (hunk 8 above)
Comment (amanCypherock) [reply src/wallet/wallet_list.h:57]: Also brief comment is same for both APIs here, which is very confusing

File: src/card_flows/card_flow_reconstruct_wallet.c [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+card_error_type_e card_flow_reconstruct_wallet(uint8_t threshold) {
+  // TODO: Support this for threshold number of cards
+
+  card_fetch_share_cfg_t configuration = {.heading = ui_text_tap_1_2_cards,
+                                          .message = ui_text_place_card_below,
+                                          .skip_card_removal = false,
```
Comment (amanCypherock) [inline src/card_flows/card_flow_reconstruct_wallet.c:104]: This should be true, as we want to skip the card removal wait.

File: src/wallet/reconstruct_seed_flow.c [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+         */
+        next_state = COMPLETED_WITH_ERRORS;
+      }
+
+      break;
+    }
+
+    case RECONSTRUCT_SEED: {
+      instruction_scr_init("", NULL);
+      instruction_scr_change_text(ui_text_processing, true);
```
Comment (amanCypherock) [inline src/wallet/reconstruct_seed_flow.c:219]: Directly call 
```suggestion
      instruction_scr_init(ui_text_processing, NULL);
```

File: src/wallet/reconstruct_seed_flow.c [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+          get_state_on_input_scr(PASSPHRASE_CONFIRM, EARLY_EXIT, TIMED_OUT);
+      break;
+    }
+
+    case PASSPHRASE_CONFIRM: {
+      char display[65];
+      snprintf(
+          display, sizeof(display), "%s", flow_level.screen_input.input_text);
+      address_scr_init(ui_text_confirm_passphrase, display, false);
+      memzero(display, sizeof(display));
```
Comment (amanCypherock) [inline src/wallet/reconstruct_seed_flow.c:154]: We could remove the variable display here

File: src/card_flows/card_flow_reconstruct_wallet.c (hunk 10 above)
Comment (kartikCypherock) [reply src/card_flows/card_flow_reconstruct_wallet.c:104]: Ok, updated

File: src/wallet/reconstruct_seed_flow.c (hunk 11 above)
Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.c:219]: Updated

File: src/wallet/reconstruct_seed_flow.c (hunk 12 above)
Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.c:154]: How can we do that? Passphrase is stored in that variable to display using `address_scr_init` in line #153

File: src/wallet/wallet_list.h (hunk 9 above)
Comment (kartikCypherock) [reply src/wallet/wallet_list.h:72]: This function is for the coin application to get the name of the wallet.. Name of the wallet is required to trigger user confirmation before any operation on the wallet

Comment (kartikCypherock) [reply src/wallet/wallet_list.h]: I have updated the name to `get_wallet_name_by_id`

File: src/wallet/wallet_list.h (hunk 8 above)
Comment (kartikCypherock) [reply src/wallet/wallet_list.h]: Updated name to `get_wallet_data_by_id`

File: src/wallet/reconstruct_seed_flow.c (hunk 12 above)
Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.c]: Updated to use scroll page

**Comment by amanCypherock** [comment]
As there have been multiple changes in this PR, we should test the change first before merging.

File: src/wallet/reconstruct_seed_flow.h [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+ * reconstruction.
+ *
+ * @param wallet_id The wallet_id of the wallet which needs to be reconstructed
+ * @param seed_out Pointer to buffer where the seed will be copied after
+ * regeneration
+ * @return true If the reconstruction flow completed successfully and buffer
+ * pointed by seed_out is filled with the seed
+ * @return false If the reconstruction flow could not be completed
+ */
+bool reconstruct_seed_flow(const uint8_t *wallet_id, uint8_t *const seed_out);
```
Comment (amanCypherock) [inline src/wallet/reconstruct_seed_flow.h:48]: The declaration and definition of this function are different, please make this consistent.

Comment (kartikCypherock) [reply src/wallet/reconstruct_seed_flow.h]: Good catch, fixed the mismatch

**Comment by kartikCypherock** [comment]
> As there have been multiple changes in this PR, we should test the change first before merging.

Hi Aman, rightly pointed out, I have tested the changes and it works well. We can move with merging.

**Review by amanCypherock — APPROVED** [review]

**Pull request merged**